src/main/java/io/numaproj/numaflow/function/FunctionServer.java

```diff
@@ -69,8 +70,8 @@ public FunctionServer registerMapperT(MapTHandler mapTHandler) {
         return this;
     }
 
-    public FunctionServer registerReducer(Class<? extends Reducer> groupByClass) {
-        this.functionService.setReduceHandler(groupByClass);
+    public FunctionServer registerReducer(ReducerFactory<? extends Reducer> reducerFactory) {
```

Comment:
The method name stays `registerReducer`, why not keep it as it is and pass in a Reducer instance?

```
public FunctionServer registerReducer(Reducer reducer)
```

and invoke it using

```
new FunctionServer().registerReducer(new SumFactory().createReducer()).start();
```

Reply:
Please look at the FunctionService code, we create a new Reducer for each key. So we can't accept a reducer object as is, instead we accept a factory that creates the Reducer and use that factory to create an object of Reducer for each key combination.

Reply:
I will change the method name to `resigerReducerFactory`

Reply:
> We create a new Reducer for each key.

Is it at line 94 of class `ReduceSupervisorActor`? For each different key, we create a brand new actor. For these new actors, are they using different Reducer implementations?

> use that factory to create an object of Reducer for each key combination.

Can multiple keys/actors share the same Reducer object?

Reply:
Adding a note for my future reference: For each different keys, we shouldn't share the same Reducer object. Taking EvenOddCounter reducer as example, each reducer object has its own local `sum` instance variable, if even key and odd key share the same reducer object, they will be sharing the same `sum` variable, which completely breaks the purpose of reducer(aggregation by key).